src/data/history.ts

```diff
@@ -212,65 +268,68 @@ const equalState = (obj1: LineChartState, obj2: LineChartState) =>
 const processTimelineEntity = (
   localize: LocalizeFunc,
   language: FrontendLocaleData,
-  states: HassEntity[]
+  entityId: string,
+  states: EntityHistoryState[]
 ): TimelineEntity => {
   const data: TimelineState[] = [];
-  const last_element = states.length - 1;
-
+  const last: EntityHistoryState = states[states.length - 1];
   for (const state of states) {
-    if (data.length > 0 && state.state === data[data.length - 1].state) {
+    if (data.length > 0 && state.s === data[data.length - 1].state) {
       continue;
     }
-
-    // Copy the data from the last element as its the newest
-    // and is only needed to localize the data
-    if (!state.entity_id) {
-      state.attributes = states[last_element].attributes;
-      state.entity_id = states[last_element].entity_id;
-    }
-
     data.push({
-      state_localize: computeStateDisplay(localize, state, language),
-      state: state.state,
-      last_changed: state.last_changed,
+      state_localize: computeStateDisplayFromEntityAttributes(
+        localize,
+        language,
+        entityId,
+        state.a || last.a,
+        state.s
+      ),
+      state: state.s,
+      last_changed: state.lc * 1000,
     });
   }
 
   return {
-    name: computeStateName(states[0]),
-    entity_id: states[0].entity_id,
+    name: computeStateNameFromEntityAttributes(entityId, states[0].a),
+    entity_id: entityId,
     data,
   };
 };
 
 const processLineChartEntities = (
   unit,
-  entities: HassEntity[][]
+  entities: HistoryStates
 ): LineChartUnit => {
   const data: LineChartEntity[] = [];
 
-  for (const states of entities) {
-    const last: HassEntity = states[states.length - 1];
-    const domain = computeStateDomain(last);
+  Object.keys(entities).forEach((entityId) => {
+    const states = entities[entityId];
+    const last: EntityHistoryState = states[states.length - 1];
+    const domain = computeDomain(entityId);
     const processedStates: LineChartState[] = [];
 
     for (const state of states) {
       let processedState: LineChartState;
 
       if (DOMAINS_USE_LAST_UPDATED.includes(domain)) {
         processedState = {
-          state: state.state,
-          last_changed: state.last_updated,
+          state: state.s,
+          last_changed: state.lu * 1000,
           attributes: {},
         };
 
         for (const attr of LINE_ATTRIBUTES_TO_KEEP) {
-          if (attr in state.attributes) {
-            processedState.attributes![attr] = state.attributes[attr];
+          if (attr in state.a) {
+            processedState.attributes![attr] = state.a[attr];
           }
         }
       } else {
-        processedState = state;
+        processedState = {
+          state: state.s,
+          last_changed: state.lc * 1000,
+          attributes: {},
+        };
```

Comment:
I could implement the fallback we discussed here or wait for a followup PR?

https://github.com/home-assistant/core/pull/71843#issuecomment-1126852642

Reply:
https://github.com/home-assistant/core/pull/71886 Created this one.

Reply:
Actually it was just as easy to make the code backwards compatible so I did that.